### PR TITLE
Feature/Events & planning

### PR DESCRIPTION
**Pull request by Yoronex**
Adds events and corresponding plannings to SudoSOS.

Events can be many things, like weekly borrels, extended weekly borrels, other GEWIS activities where BAC people are needed, activities from external parties where the BAC is needed, etc. For all these events, the BAC can create a new event. When creating events, shifts are added, which have to be done by one or more people. Shifts are based on roles: whoever has a shift's role, can fill in their availability. These "answer sheets" are synchronised every night in case users get added to a group, or are removed from a group. However, they are only removed once the event has taken place. For every answer sheet, the BAC can choose who is selected for the shift. This can be one or more people.

Shifts have to be added manually and should be reused between events. If a shift is used at least once in an event and it is deleted, it will be soft deleted. 

Future extensions for this PR are:
- Transactions can be linked to events. In a "borrelmode" POS when an event is almost starting, the user can choose to link all transactions to the event. This opens many opportunities for amazing data processing, like comparing borrels.
- Know how many Aquarius is bought a day after a borrel.
- Automatically enable and disable SudoSOS posters for the poster screens (you know when a borrel takes place).
- https://isdeborrelverlengd.nl

### CODE REVIEW

**Comment by Yoronex** [comment]
Added an extra endpoint, namely to get an overview of how many times someone has been selected for a shift, optionally between a certain time interval and/or for events for a certain type.

**Review by JustSamuel — APPROVED** [review]
Small questions, but looks good IMHO

File: src/entity/event/event.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+} from 'typeorm';
+import BaseEntity from '../base-entity';
+import User from '../user/user';
+import EventShiftAnswer from './event-shift-answer';
+import EventShift from './event-shift';
+
+export enum EventType {
+  BORREL = 'BORREL', // Weekly GEWIS borrel, both normal borrels and extended borrels
+  EXTERNAL_BORREL = 'EXTERNAL_BORREL', // Borrel with/for external party
+  OTHER = 'OTHER', // All other activities
```
[the remark this answers is not in the document]
Comment (Yoronex) [reply src/entity/event/event.ts:29]: Dit is een bewuste keuze. Zie bijvoorbeeld endpoint `GET /eventshifts/{id}/count`. Je geeft een optionele `eventType` om te kunnen filteren tussen bepaalde type events. Ik heb overwogen om deze attribute simpelweg een string te maken, maar dan is het heel makkelijk om "fouten" te maken, bijvoorbeeld met een typfout. Ook heb je dan een endpoint nodig om alle reeds bestaande type events op te halen.

File: src/controller/event-controller.ts [hunk 2]
```diff
@@ -0,0 +1,441 @@
...
+    }
+  }
+
+  /**
+   * Create an event with its corresponding answers objects
+   * @route POST /events
+   * @group events - Operations of the event controller
+   * @operationId createEvent
+   * @security JWT
+   * @param {CreateEventRequest.model} body.body.required
```
[the remark this answers is not in the document]
Comment (Yoronex) [reply src/controller/event-controller.ts:192]: Hoeft niet, want nu heet het veld simpelweg "body" en is het van het type "body". Vind je dat vervelend? Want dan moeten we dat aanpassen.

image

**Pull request merged**